examples/render-svg.rs

```diff
@@ -0,0 +1,440 @@
+//!
+//! For each feature, we retrieve the coordinates lazily from osm nodes, and
+//! then produce polylines styled based on the category, cf. `render_svg`
+//! function. The coordinates are in lon, lat.
+//!
+//! Inside of svg we just use the coordinates as is (except for swapped x/y
+//! axes), plus we apply a transformation to adjust the coordinates to the
+//! viewport. Obviously, it is slower the render such svg on the screen.
+//! However, the final svg contains already so many polyline, that having alrady
+//! transformed coordinates does not change much. If you need speed when showing
+//! the svg, feel free to apply simplifications in this program.
+
+extern crate flatdata;
+extern crate itertools;
+extern crate osmflat;
+#[macro_use]
+extern crate smallvec;
+extern crate structopt;
+extern crate svg;
+
+use flatdata::{Archive, FileResourceStorage};
+use smallvec::SmallVec;
+use std::fmt::Write;
+use structopt::StructOpt;
+use svg::node::element;
+use svg::Document;
+
+use std::f64;
+use std::io;
+use std::ops::Range;
+use std::path::PathBuf;
+use std::str;
+
+/// Helper function to read a string from osmflat.
+fn substring(strings: &str, start: u64) -> &str {
+    let start = start as usize;
+    let end = strings[start..].find('\0').expect("invalid string");
+    &strings[start..start + end]
+}
+
+/// Geographic coordinates represented by (latitude, longitude).
+#[derive(Debug, Clone, Copy, Default, PartialEq, PartialOrd)]
+struct GeoCoord {
+    lat: f64,
+    lon: f64,
+}
+
+impl GeoCoord {
+    fn min(self, other: Self) -> Self {
+        Self {
+            lat: self.lat.min(other.lat),
+            lon: self.lon.min(other.lon),
+        }
+    }
+
+    fn max(self, other: Self) -> Self {
+        Self {
+            lat: self.lat.max(other.lat),
+            lon: self.lon.max(other.lon),
+        }
+    }
+}
+
+/// Convert osmflat Node into GeoCoord.
+impl<'a> From<osmflat::RefNode<'a>> for GeoCoord {
+    fn from(node: osmflat::RefNode<'a>) -> Self {
+        const COORD_SCALE: f64 = 0.000_000_001;
+        Self {
+            lat: node.lat() as f64 * COORD_SCALE,
+            lon: node.lon() as f64 * COORD_SCALE,
+        }
+    }
+}
+
+/// Polyline which can be transformed into an iterator over `GeoCoord`'s.
+struct Polyline {
+    inner: SmallVec<[Range<u64>; 4]>,
+}
+
+impl From<Range<u64>> for Polyline {
+    fn from(range: Range<u64>) -> Self {
+        Self {
+            inner: smallvec![range],
+        }
+    }
+}
+
+impl Polyline {
+    fn into_iter(self, archive: osmflat::Osm) -> PolylineIter {
+        let next_element = self.inner.first().map(|range| range.start).unwrap_or(0);
+        PolylineIter {
+            archive,
+            inner: self.inner,
+            next_range: 0,
+            next_element,
+        }
+    }
+}
+
+/// Iterator over osmflat nodes.
+///
+/// A polyline contains ranges of nodes. This iterator iterates over the ranges
+/// and resolves nodes as GeoCoord's.
+struct PolylineIter {
```

Comment:
Could you replace this iterator with a map of flattened ranges? (e.g. just a function that creates you the transform). Could be much more concise.